rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/delay/TbMsgDelayNode.java

```diff
@@ -23,61 +23,194 @@
+        nodeDescription = "Delays incoming message",
         nodeDetails = "Delays messages for a configurable period. " +
                 "Please note, this node acknowledges the message from the current queue (message will be removed from queue). " +
-                "Deprecated because the acknowledged message still stays in memory (to be delayed) and this " +
-                "does not guarantee that message will be processed even if the \"retry failures and timeouts\" processing strategy will be chosen.",
+                "The acknowledged message stays in memory to be delayed. Additionally, if the cache type is set to <b>Redis</b>, " +
+                "the message will be persisted in the cache, which guarantees that message will be processed even after the server restart." +
+                "<b>Important note:</b> If the incoming message is processed in the queue with a sequential processing strategy configured, " +
+                "the message acknowledgment that used in the rule node logic will trigger the next message to be processed by the queue.",
         icon = "pause",
         uiResources = {"static/rulenode/rulenode-core-config.js"},
         configDirective = "tbActionNodeMsgDelayConfig"
 )
 public class TbMsgDelayNode implements TbNode {
 
     private static final String TB_MSG_DELAY_NODE_MSG = "TbMsgDelayNodeMsg";
+    private static final String DELAYED_ORIGINATOR_IDS_CACHE_KEY = "delayed_originator_ids";
+    private static final TbMsgMetaData EMPTY_META_DATA = new TbMsgMetaData();
 
     private TbMsgDelayNodeConfiguration config;
-    private Map<UUID, TbMsg> pendingMsgs;
+    private final Map<Integer, Set<EntityId>> partitionsEntityIdsMap;
+    private final Map<EntityId, Map<UUID, TbMsg>> entityIdPendingMsgs;
+
+    public TbMsgDelayNode() {
+        this.partitionsEntityIdsMap = new HashMap<>();
+        this.entityIdPendingMsgs = new HashMap<>();
+    }
 
     @Override
     public void init(TbContext ctx, TbNodeConfiguration configuration) throws TbNodeException {
         this.config = TbNodeUtils.convert(configuration, TbMsgDelayNodeConfiguration.class);
-        this.pendingMsgs = new HashMap<>();
+        getPendingMsgsFromCacheAndSchedule(ctx);
     }
 
     @Override
     public void onMsg(TbContext ctx, TbMsg msg) {
         if (msg.getType().equals(TB_MSG_DELAY_NODE_MSG)) {
-            TbMsg pendingMsg = pendingMsgs.remove(UUID.fromString(msg.getData()));
-            if (pendingMsg != null) {
-                ctx.enqueueForTellNext(pendingMsg, SUCCESS);
+            processDelayMsg(ctx, msg);
+        } else {
+            processOnRegularMsg(ctx, msg);
+        }
+    }
+
+    private void getPendingMsgsFromCacheAndSchedule(TbContext ctx) {
+        log.trace("[{}] Going to fetch delayed messages from cache ...", ctx.getSelfId());
+        Set<EntityId> entityIds = ctx.getRuleNodeCacheService().getEntityIds(DELAYED_ORIGINATOR_IDS_CACHE_KEY);
```

Comment:
This should be pageable with pagination of 1000 values or similar. Big environments may fail by OOM.

Reply:
This code block should be refactored to a static utility or abstract method. It will be the same for all rule nodes except the block that is processing `pendingMsgs`

Reply:
We should use it in the 'delay', 'deduplication', and maybe even analytics nodes.